apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -1369,6 +1186,157 @@ handle_info(Info, StateName, StateData) ->
     ?ERROR_MSG("Unexpected info: ~p", [Info]),
     fsm_next_state(StateName, StateData).
 
+handle_routed(<<"presence">>, From, To, Packet, StateData) ->
+    handle_routed_presence(From, To, Packet, StateData);
+handle_routed(<<"broadcast">>, _From, _To, Packet, StateData) ->
+    handle_routed_broadcast(Packet, StateData);
+handle_routed(<<"iq">>, From, To, Packet, StateData) ->
+    handle_routed_iq(From, To, Packet, StateData);
+handle_routed(<<"message">>, From, To, Packet, StateData) ->
+    case privacy_check_packet(StateData, From, To, Packet, in) of
+	allow ->
+	    {true, Packet#xmlel.attrs, StateData};
+	deny ->
+	    {false, Packet#xmlel.attrs, StateData}
+    end;
+handle_routed(_, _From, _To, Packet, StateData) ->
+    {true, Packet#xmlel.attrs, StateData}.
+
+handle_routed_iq(From, To, Packet = #xmlel{attrs = Attrs}, StateData) ->
+    case jlib:iq_query_info(Packet) of
+	%% TODO: Support for mod_last / XEP-0012. Can we move it to the respective module?
```

Comment:
I'm not sure yet, but after a quick glance it seems `StateData` here is only used for checking whether `From` is subscribed to *our* presence (which can be told outside of C2S by looking at the roster) and privacy list check. This would be more costly when done from `mod_last` (extra roundtrips to the DB / roster table), but would slim c2s down. If necessary a new hook is OK in my opinion and doesn't introduce the extra overhead.